Adding `-arch=sm_XY` to kokkos link line creates trouble for apps that use f90 linker

I''m using the latest release, 3.2.00, and am having trouble linking kokkos to our application, due to the `-arch=..` flag. Our application has a F90 main entry, so we use a f90 linker. This causes the flag `-arch=sm_*` to be unrecognized.

The problem seems to stem from the following line:

```
IF(KOKKOS_ENABLE_CUDA_RELOCATABLE_DEVICE_CODE OR KOKKOS_CXX_COMPILER_ID STREQUAL NVIDIA)
    GLOBAL_APPEND(KOKKOS_LINK_OPTIONS "${CUDA_ARCH_FLAG}=${FLAG}")
ENDIF()

```
I have a couple of questions on this:
1) if the user enabled cuda's rdc, then isn't the compiler always going to be nvidia? In which case, what's the point of checking the compiler id?
2) I understand why one would need that flag for rdc, but is it really needed always? Meaning, if I don't use rdc, do I really need the arch flag on my link line? I manually removed it from the `link.txt` for my f90 executable target, and it compiled and run. Maybe it was just a coincidence. But if not, then shouldn't kokkos add this flag only to the _private_ section (in cmake lingo) of the target's link properties?

Edit: [this](https://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/index.html#options-for-altering-compiler-linker-behavior-dlink-time-opt) seems to suggest this is needed for some link time optimization. Assuming that's the case (and point 2 above doesn't apply), is there a way to disable this optimization, and not get the flag in the link options?

Sure:

CMakeLists.txt:
```
project(reproducer CXX Fortran)

add_subdirectory(${KOKKOS_PATH} ${CMAKE_CURRENT_BINARY_DIR}/kokkos)

add_executable(test main.f90)
target_link_libraries(test kokkos)
set_target_properties(test PROPERTIES LINKER_LANGUAGE Fortran)
```
where you need to set `KOKKOS_PATH` to kokkos src folder (I used the 3.2.00 release tag).

main.f90:
```
program test

  integer :: i

  i = 1 

  print*, "i: ", i

end program test
```
I used gcc 7.2, and cuda 10.1.105. Result:

```
...
[ 92%] Built target kokkoscontainers
[ 96%] Building Fortran object CMakeFiles/test.dir/main.o
[100%] Linking Fortran executable test
gfortran: error: unrecognized command line option ‘-arch=sm_70’
make[2]: *** [test] Error 1
make[1]: *** [CMakeFiles/test.dir/all] Error 2
make: *** [all] Error 2
```

--- a/BUILD.md
+++ b/BUILD.md
@@ -52,12 +52,16 @@ There are numerous device backends, options, and architecture-specific optimizat
 ````
 which activates the OpenMP backend. All of the options controlling device backends, options, architectures, and third-party libraries (TPLs) are given below.
 
-## Platform-specific Problems
+## Known Issues<a name="KnownIssues"></a>
 
 ### Cray
 
 * The Cray compiler wrappers do static linking by default. This seems to break the Kokkos build. You will likely need to set the environment variable `CRAYPE_LINK_TYPE=dynamic` in order to link correctly. Kokkos warns during configure if this is missing.
 * The Cray compiler identifies to CMake as Clang, but it sometimes has its own flags that differ from Clang. We try to include all exceptions, but flag errors may occur in which a Clang-specific flag is passed that the Cray compiler does not recognize.
+
+### Fortran
+
+* In a mixed C++/Fortran code, CMake will use the C++ linker by default. If you override this behavior and use Fortran as the link language, the link may break because Kokkos adds linker flags expecting the linker to be C++. Prior to CMake 3.18, Kokkos has no way of detecting in downstream projects that the linker was changed to Fortran.  From CMake 3.18, Kokkos can use generator expressions to avoid adding flags when the linker is not C++. Note: Kokkos will not add any linker flags in this Fortran case. The user will be entirely on their own to add the appropriate linker flags.
 
 ## Spack
 An alternative to manually building with the CMake is to use the Spack package manager.

--- a/README.md
+++ b/README.md
@@ -100,6 +100,10 @@ For specifics see the LICENSE file contained in the repository or distribution.
 * ARM
     * Pthreads backend
 
+### Build system:
+* CMake >= 3.10: required
+* CMake >= 3.13: recommended
+* CMake >= 3.18: Fortran linkage. This does not affect most mixed Fortran/Kokkos builds. See [build issues](BUILD.md#KnownIssues).
 
 Primary tested compiler are passing in release mode
 with warnings as errors. They also are tested with a comprehensive set of

--- a/cmake/kokkos_tribits.cmake
+++ b/cmake/kokkos_tribits.cmake
@@ -238,12 +238,21 @@ FUNCTION(KOKKOS_SET_LIBRARY_PROPERTIES LIBRARY_NAME)
     ""
     ${ARGN})
 
-  IF(${CMAKE_VERSION} VERSION_GREATER_EQUAL "3.13")
-    #great, this works the "right" way
+  IF(${CMAKE_VERSION} VERSION_GREATER_EQUAL "3.18")
+    #I can use link options
+    #check for CXX linkage using the simple 3.18 way
+    TARGET_LINK_OPTIONS(
+      ${LIBRARY_NAME} PUBLIC
+      $<$<LINK_LANGUAGE:CXX>:${KOKKOS_LINK_OPTIONS}>
+    )
+  ELSEIF(${CMAKE_VERSION} VERSION_GREATER_EQUAL "3.13")
+    #I can use link options
+    #just assume CXX linkage
     TARGET_LINK_OPTIONS(
       ${LIBRARY_NAME} PUBLIC ${KOKKOS_LINK_OPTIONS}
     )
   ELSE()
+    #assume CXX linkage, we have no good way to check otherwise
     IF (PARSE_PLAIN_STYLE)
       TARGET_LINK_LIBRARIES(
         ${LIBRARY_NAME} ${KOKKOS_LINK_OPTIONS}

